2006-12-19  Jakub Jelinek  <jakub@redhat.com>

	* stdlib/Makefile (tests): Add tst-makecontext.
	* stdlib/tst-makecontext.c: New test.

	* sysdeps/unix/sysv/linux/powerpc/powerpc32/makecontext.S
	(__makecontext): Don't realign uc_mcontext.uc_regs.

2006-12-13  Jakub Jelinek  <jakub@redhat.com>

	* sysdeps/unix/sysv/linux/powerpc/powerpc32/setcontext.S: Include
	kernel-features.h.

--- libc/sysdeps/unix/sysv/linux/powerpc/powerpc32/setcontext.S	29 Dec 2005 21:03:38 -0000	1.12
+++ libc/sysdeps/unix/sysv/linux/powerpc/powerpc32/setcontext.S	13 Dec 2006 09:09:24 -0000	1.13
@@ -20,6 +20,7 @@
 #include <sysdep.h>
 #include <rtld-global-offsets.h>
 #include <shlib-compat.h>
+#include <kernel-features.h>
 
 #define __ASSEMBLY__
 #include <asm/ptrace.h>
--- libc/sysdeps/unix/sysv/linux/powerpc/powerpc32/makecontext.S	7 Jan 2006 03:51:11 -0000	1.6
+++ libc/sysdeps/unix/sysv/linux/powerpc/powerpc32/makecontext.S	19 Dec 2006 15:33:10 -0000	1.7
@@ -26,9 +26,7 @@
 
 ENTRY(__makecontext)
 	/* Set up the first 7 args to the function in its registers */
-	addi	r11,r3,_UC_REG_SPACE+12
-	clrrwi  r11,r11,4
-	stw	r11,_UC_REGS_PTR(r3)
+	lwz	r11,_UC_REGS_PTR(r3)
 	stw	r6,_UC_GREGS+(PT_R3*4)(r11)
 	stw	r7,_UC_GREGS+(PT_R4*4)(r11)
 	stw	r8,_UC_GREGS+(PT_R5*4)(r11)
--- libc/stdlib/Makefile	11 Dec 2006 21:43:10 -0000	1.113
+++ libc/stdlib/Makefile	19 Dec 2006 15:34:08 -0000	1.115
@@ -67,7 +67,8 @@ tests		:= tst-strtol tst-strtod testmb t
 		   tst-xpg-basename tst-random tst-random2 tst-bsearch	    \
 		   tst-limits tst-rand48 bug-strtod tst-setcontext	    \
 		   test-a64l tst-qsort tst-system testmb2 bug-strtod2	    \
-		   tst-atof1 tst-atof2 tst-strtod2 tst-strtod3 tst-rand48-2
+		   tst-atof1 tst-atof2 tst-strtod2 tst-strtod3 tst-rand48-2 \
+		   tst-makecontext
 
 include ../Makeconfig
 
--- libc/stdlib/tst-makecontext.c	1 Jan 1970 00:00:00 -0000
+++ libc/stdlib/tst-makecontext.c	19 Dec 2006 15:33:58 -0000	1.1
@@ -0,0 +1,57 @@
+/* Copyright (C) 2006 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, write to the Free
+   Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
+   02111-1307 USA.  */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <ucontext.h>
+
+ucontext_t ucp;
+char st1[8192];
+__thread int thr;
+
+void
+cf (int i)
+{
+  if (i != 78 || thr != 94)
+    {
+      printf ("i %d thr %d\n", i, thr);
+      exit (1);
+    }
+  exit (0);
+}
+
+int
+main (void)
+{
+  if (getcontext (&ucp) != 0)
+    {
+      puts ("getcontext failed");
+      return 1;
+    }
+  thr = 94;
+  ucp.uc_link = NULL;
+  ucp.uc_stack.ss_sp = st1;
+  ucp.uc_stack.ss_size = sizeof st1;
+  makecontext (&ucp, (void (*) ()) cf, 1, 78);
+  if (setcontext (&ucp) != 0)
+    {
+      puts ("setcontext failed");
+      return 1;
+    }
+  return 2;
+}
